presto-main/src/main/java/com/facebook/presto/sql/planner/TranslationMap.java

```diff
@@ -50,6 +51,9 @@
     private final Map<Expression, Symbol> expressionToSymbols = new HashMap<>();
     private final Map<Expression, Expression> expressionToExpressions = new HashMap<>();
 
+    // memoize translated expressions so that the same output expression will be produced for each input expression
+    private final IdentityHashMap<Expression, Expression> translatedExpressions = new IdentityHashMap<>();
```

Comment:
Is this memoization required for correctness, or is it a performance optimization?